src/dict.h

```diff
@@ -55,8 +55,8 @@ typedef struct dictType {
     uint64_t (*hashFunction)(const void *key);
     void *(*keyDup)(dict *d, const void *key);
     int (*keyCompare)(dict *d, const void *key1, const void *key2);
-    void (*keyDestructor)(dict *d, void *key);
-    void (*valDestructor)(dict *d, void *obj);
+    void (*keyDestructor)(struct dictType *type, void *key);
+    void (*valDestructor)(struct dictType *type, void *obj);
```

Comment:
I'm not sure why we even need a `dictType` passed to these routines?  The purpose of the callback is to dispose of the key or value.  So the callback function will be unique to the key or value - not related to the `dict` or `dictType`.

If there needs to be a difference in how the callback functions based on the `dictType`, then it should be a different callback - which is configured into the `dictType`.

This callback interface is C-style polymorphism.  With polymorphism, you don't usually have IF logic inside a function.  Instead, you have 2 functions.

Reply:
Some callbacks are passed the dict `d` because they want to access the dict metadata inside the dict. Examples are dictRehashingStarted and dictRehashingComplete used by kvstore, which stores a pointer to the kvstore instance in the dict metadata.

So I can see why a destructor potentially could want to be passed a dict `d`, but I can't see why any of these callback would need the dictType. But I agree, the destructors in our code base don't need any of these, so let's just pass the object to be freed.